src/content/en/fundamentals/engage-and-retain/push-notifications/web-push-protocol.md

```diff
@@ -0,0 +1,613 @@
+
+So our Pseudo Random Key is simply running the auth, shared secret and a piece of encoding info
+through HKDF (i.e. making it cryptographically stronger).
+
+### Context
+
+The "context" is a set of bytes that is used to calculate two values later on in the encryption
+browser. It's essentially an array of bytes containing the subscription public key and the
+local public key.
+
+    const keyLabel = new Buffer('P-256\0', 'utf8');
+
+    // Convert subscription public key into a buffer.
+    const subscriptionPubKey = new Buffer(subscription.keys.p256dh, 'base64');
+
+    const subscriptionPubKeyLength = new Uint8Array(2);
+    subscriptionPubKeyLength[0] = 0;
+    subscriptionPubKeyLength[1] = subscriptionPubKey.length;
+
+    const localPublicKeyLength = new Uint8Array(2);
+    subscriptionPubKeyLength[0] = 0;
+    subscriptionPubKeyLength[1] = localPublicKey.length;
+
+    const contextBuffer = Buffer.concat([
+      keyLabel,
+      subscriptionPubKeyLength.buffer,
+      subscriptionPubKey,
+      localPublicKeyLength.buffer,
+      localPublicKey,
+    ]);
+    
+The final context buffer is a label, the number of bytes in the subscription public key,
+followed by the key itself, then the number of bytes local public key, followed by the key
+itself.
+
+With this context value we can use it in the creation of a nonce and a content encryption key
+(CEK).
+
+### Content Encryption Key and Nonce
+
+A [nonce](https://en.wikipedia.org/wiki/Cryptographic_nonce) is a value that prevents replay
+attacks as it should only be used once.
+
+The content encryption key (CEK) is the key that will ultimately be used to encrypt our payload.
+
+First we need to create the bytes of data for the nonce and CEK, which is simply a content
+encoding string followed by the context buffer we just calculated:
+
+    const nonceEncBuffer = new Buffer('Content-Encoding: nonce\0', 'utf8');
+    const nonceInfo = Buffer.concat([nonceEncBuffer, contextBuffer]);
+
+    const cekEncBuffer = new Buffer('Content-Encoding: aesgcm\0');
+    const cekInfo = Buffer.concat([cekEncBuffer, contextBuffer]);
+    
+This information is run through HKDF combining the salt and PRK with the nonceInfo and cekInfo:
+
+    // The nonce should be 12 bytes long
+    const nonce = hkdf(salt, prk, nonceInfo, 12);
+
+    // The CEK should be 16 bytes long
+    const contentEncryptionKey = hkdf(salt, prk, cekInfo, 16);
+    
+This gives us our nonce and content encryption key.
+
+### Perform the Encryption
+
+Now that we have our content encryption key, we can encrypt the payload.
+
+We create an AES128 Cipher using the content encryption key
+as the key and the nonce is as an initialization vector.
+
+In Node this is done like so:
+
+    const cipher = crypto.createCipheriv(
+      'id-aes128-GCM', contentEncryptionKey, nonce);
+    
+Before we encrypt our payload, we need to define how much padding we wish to
+add to the front of the payload. The reason you'd want to add padding is that
```

Comment:
Change to 'The reason *we'd* want'. Make sure your pronouns are consistent throughout.